File: src/main/java/org/pyjjs/scheduler/core/api/impl/actors/task/TaskActorState.java
package org.pyjjs.scheduler.core.api.impl.actors.task;

import akka.actor.ActorContext;
import org.pyjjs.scheduler.core.api.impl.actors.common.ActorState;
import org.pyjjs.scheduler.core.api.impl.actors.common.SourceBasedActorState;
import org.pyjjs.scheduler.core.api.impl.actors.resource.messages.OfferMessage;
import org.pyjjs.scheduler.core.model.Task;

import java.util.HashSet;
import java.util.Set;

public class TaskActorState extends SourceBasedActorState<Task> {

    private Set<OfferMessage> offers = new HashSet<>();
    private boolean checkOffersScheduled = false;
    private Double discontent = null;

    public TaskActorState(ActorContext actorContext) {
        super(actorContext);
    }

    public void addOffer(OfferMessage resourcePrice) {
        offers.add(resourcePrice);
    }

    public Set<OfferMessage> getOffers() {
        return offers;
    }

    public boolean checkOffersIsScheduled() {
        return checkOffersScheduled;
    }

    public void setCheckOffersScheduled(boolean checkOffersScheduled) {
        this.checkOffersScheduled = checkOffersScheduled;
    }

    public Double getDiscontent() {
        return discontent;
    }

    public void setDiscontent(Double discontent) {
        this.discontent = discontent;
    }

    @Override
    protected ActorState copySelf() {
        return this;
    }
}

File: src/main/kotlin/org/pyjjs/scheduler/core/api/impl/actors/task/TaskActorState.kt
package org.pyjjs.scheduler.core.api.impl.actors.task

import akka.actor.ActorContext
import akka.actor.ActorRef
import org.pyjjs.scheduler.core.api.impl.actors.common.ActorState
import org.pyjjs.scheduler.core.api.impl.actors.common.SourceBasedActorState
import org.pyjjs.scheduler.core.api.impl.actors.common.messages.ResourceHasFullPlacementMessage
import org.pyjjs.scheduler.core.api.impl.actors.resource.messages.OfferMessage
import org.pyjjs.scheduler.core.model.Task
import org.pyjjs.scheduler.core.placement.Placement
import java.util.*

class TaskActorState(actorContext: ActorContext) : SourceBasedActorState<Task>(actorContext) {

    enum class Status { UNPLACED, WAIT_OFFER, WAIT_RESERVING, PLACED }

    private val offers = HashSet<OfferMessage>()
    private var checkOffersScheduled = false
    var discontent: Double? = null
    var status: Status = Status.UNPLACED
    var placementMessageQueue = PriorityQueue<ResourceHasFullPlacementMessage>()
    var offersById = HashMap<UUID, ResourceHasFullPlacementMessage>()

    //messaging info
    private val notMatchedResources = HashSet<ActorRef>()
    private val resourcesWithoutFreeTimes = HashSet<ActorRef>()

    fun addOffer(resourcePrice: OfferMessage) {
        offers.add(resourcePrice)
    }

    fun getOffers(): Set<OfferMessage> {
        return offers
    }

    fun addNotMatchedResource(ref: ActorRef) {
        notMatchedResources.add(ref)
    }

    fun addResourceWithoutFreeTimes(ref: ActorRef) {
        resourcesWithoutFreeTimes.add(ref)
    }

    fun checkOffersIsScheduled(): Boolean {
        return checkOffersScheduled
    }

    fun setCheckOffersScheduled(checkOffersScheduled: Boolean) {
        this.checkOffersScheduled = checkOffersScheduled
    }

    override fun copySelf() = this
}
